fix: update readme and contributing docs for new standards (#365)
aws-samples/aws-cdk-examples: CONTRIBUTING.md

Please read through this document before submitting any issues or pull requests to ensure we have all the necessary
information to effectively respond to your bug report or contribution.

## Table of Contents
1. [Reporting Bugs/Feature Requests](#Reporting)
2. [Contributing via Pull Requests](#Pulls)
3. [Adding an Example](#New)

> 4. [Linking External Example](#Extern)

Reviewer: The link doesn't seem to be working

1. Fork the repository.
2. Modify the source; please focus on the specific change you are contributing. If you also reformat all the code, it will be hard for us to focus on your change.

> 3. Ensure local tests pass (check the [scripts](https://github.com/aws-samples/aws-cdk-examples/tree/master/scripts) directory)

Reviewer 2: Check the scripts directory for what?

Author: ```suggestion
3. Ensure local tests pass automated tests (check the [scripts](https://github.com/aws-samples/aws-cdk-examples/tree/master/scripts) directory)
```

4. Commit to your fork using clear commit messages.
5. Send us a pull request, answering any default questions in the pull request interface.
6. Pay attention to any automated CI failures reported in the pull request, and stay involved in the conversation.

GitHub provides additional document on [forking a repository](https://help.github.com/articles/fork-a-repo/) and
[creating a pull request](https://help.github.com/articles/creating-a-pull-request/).

## Adding an Example <a name="New"></a>

> When adding a new example, there are several things important to consider as you implement (please create a `feature-request` issue to discuss these before writing your contribution):

Reviewer: Might include a link to create the issue: https://github.com/aws-samples/aws-cdk-examples/issues/new?labels=feature-request%2C+needs-triage&template=feature-request.md

Author: ```suggestion
When adding a new example, there are several things important to consider as you implement (please create a [`feature-request` issue](https://github.com/aws-samples/aws-cdk-examples/issues/new?labels=feature-request%2C+needs-triage&template=feature-request.md) to discuss these before writing your contribution):
```

### Added Value
As an official learning resource, it is important that any new examples add value to our learning resources. This means we are looking for examples that clarify additional information about developing with the CDK whether this is a common pattern, or a construct library that does not have any examples.

When considering if your example adds value, review the existing examples and think about if your example could be written with just the knowledge from the other examples.

> 

Reviewer 3: Seems like we need to be more explicit about what "added value" means to us and be transparent about how we will judge a contribution's value.

### Language Parity

> While we have historically allowed adding examples in just one of our supported languages, moving forward we would like to maintain language parity. This does not necessarily mean that you must write the example in each action yourself, but a PR will not be merged until the example has been added for each language.

Reviewer: "in each action", do you mean "each language"? Also I might, list out all of our languages as a reminder.

Author: Good catch on "each language". I am hesitant to list the languages because of the flux as we add more.

Author: ```suggestion
While we have historically allowed adding examples in just one of our supported languages, moving forward we would like to maintain language parity. This does not necessarily mean that you must write the example in each language yourself, but a PR will not be merged until the example has been added for each language.
```

If you are looking for help to achieve language parity, you can request help in either the issue related to the new example, or directly through the PR.

### Components

When adding a new example, there are several required components. The majority of these will be added when running `cdk init`.

1. General Structure

The best way to create the structure for the new example to be added is to run `npx cdk init app --language [LANGUAGE]` on an empty directory with the example name. This will generate all language-specific components needed for your new example. Please keep the suggested structure for the application (eg. typescript apps start in `./bin/example-name.ts`, and store the stack/s in `./lib/example-name-stack.ts`).

When submitting the PR, please ensure that the directory includes all relevant package manager and configuration files so that someone else viewing it only needs to run the build commands and `cdk deploy` to get the example running.

2. `README.md`

> Every example needs a comprehensive README at the root of it's directory. This README can vary depending on the example, but has several required components:

Reviewer 2: **it's** should be **its**

Author: ```suggestion
Every example needs a comprehensive README at the root of its directory. This README can vary depending on the example, but has several required components:
```

>   - Stability Banner: You can use the banner code below to indicate whether an example consists of modules that are "Stable", "Dev-Preview", or "Cfn-Only". Please reference the lowest common denominator (if all modules are stable except for one that is dev-preview, you must mark the example as dev-preview)

Reviewer: I would include a link to the Module Lifecycle guide: https://github.com/aws/aws-cdk-rfcs/blob/master/text/0107-construct-library-module-lifecycle.md to explain what "Stable", "Dev-Preview" and "Cfn-Only" mean.

Author: ```suggestion
  - Stability Banner: You can use the banner code below to indicate whether an example consists of modules that are "Stable", "Dev-Preview", or "Cfn-Only" (for more information on state, see the [Module Lifecycle guide](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0107-construct-library-module-lifecycle.md). Please reference the lowest common denominator (if all modules are stable except for one that is dev-preview, you must mark the example as dev-preview)
```

    - __Stable__
```md
<!--BEGIN STABILITY BANNER-->
---

![Stability: Stable](https://img.shields.io/badge/stability-Stable-success.svg?style=for-the-badge)

> **This is a stable example. It should successfully build out of the box**
>

> > This examples does is built on Construct Libraries marked "Stable" and does not have any infrastructure prerequisites to build.

Reviewer: "This examples does is built". Update grammar.

Author: ```suggestion
> This examples is built on Construct Libraries marked "Stable" and does not have any infrastructure prerequisites to build.
```

Reviewer: Still needs minor fix. "This example**s** is"

---
<!--END STABILITY BANNER-->
```

- __Dev Preview__
```md
<!--BEGIN STABILITY BANNER-->
---

![Stability: Developer Preview](https://img.shields.io/badge/stability-Developer--Preview-important.svg?style=for-the-badge)

> **This is an experimental example. It may not build out of the box**
>

> > This examples does is built on Construct Libraries marked "Developer Preview" and may not be updated for latest breaking changes.

Reviewer 2: **does is** -> **is**

Author: ```suggestion
> This example is built on Construct Libraries marked "Developer Preview" and may not be updated for latest breaking changes.
```

> > This examples does is built on Construct Libraries marked "Developer Preview" and may not be updated for latest breaking changes.

Reviewer: "This examples does is built". Update grammar.

---
<!--END STABILITY BANNER-->
```

  - Overview: This should be a description of the example including, but not limited to, an overview of resources being created in the example, a description of any standard/common patterns being used, as well as the overall function of the example.
  - Structure: If the example is more complex than a single stack, or the stack contains more than a couple resources, please describe the structure of application being deployed. This can be as simple as a bulleted list of resources, or as intricate as a complete infrastructure diagram.
  - Build/Deploy: Finally, the README must include the basic build/deploy instructions required for that example in said language.

## Style and Formatting <a name="Style"></a>

We strive to keep examples consistent in style and formatting. This hopefully makes navigating examples easier for users. Since the examples span different languages, we try to keep example code as idiomatic as possible.

New guidelines for various languages will be added as we define them.

### Typescript

> 1. You can use [eslint](https://eslint.org/) with the standard configuration to check for code syntax.

Reviewer 2: **check for code syntax** -> **check code syntax** or **check syntax and style** (since a linter does both)

Author: ```suggestion
1. You can use [eslint](https://eslint.org/) with the standard configuration to check code syntax and style.
```

When submitting the PR, please ensure that the directory includes all relevant package manager and configuration files so that someone else viewing it only needs to run the build commands and `cdk deploy` to get the example running.

2. `README.md`

Every example needs a comprehensive README at the root of its directory. This README can vary depending on the example, but has several required components:
  - Stability Banner: You can use the banner code below to indicate whether an example consists of modules that are "Stable", "Dev-Preview", or "Cfn-Only" (for more information on state, see the [Module Lifecycle guide](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0107-construct-library-module-lifecycle.md). Please reference the lowest common denominator (if all modules are stable except for one that is dev-preview, you must mark the example as dev-preview)

    - __Stable__
```md
<!--BEGIN STABILITY BANNER-->
---

![Stability: Stable](https://img.shields.io/badge/stability-Stable-success.svg?style=for-the-badge)

> **This is a stable example. It should successfully build out of the box**
>

> > This examples is built on Construct Libraries marked "Stable" and does not have any infrastructure prerequisites to build.

Reviewer: Typo: This examples is

Author: ```suggestion
> This example is built on Construct Libraries marked "Stable" and does not have any infrastructure prerequisites to build.
```
